### PR TITLE
Better print-out of semantics tree

### LINKED ISSUE #14669: Semantics tree print is hard to read
We are adding more and more attributes to the semantics tree so the "one line per node" format gets harder and harder to read. We should switch to a representation that sues multiple lines per node (similar to the widget and render tree).

### PR DESCRIPTION
**Pull request by goderbauer**
The semantics tree now contains a lot of information - too much to represent each node by just a single line. This PR switches the semantics tree to a multi-line-per-node representation (similar to the print-out of the render tree).

Fixes https://github.com/flutter/flutter/issues/14669

/cc @jacob314

### CODE REVIEW

File: packages/flutter/lib/src/semantics/semantics.dart [hunk 1]
```diff
@@ -1378,21 +1378,14 @@ class SemanticsNode extends AbstractNode with DiagnosticableTreeMixin {
...
-    properties.add(new FlagProperty('isTextField', value: _hasFlag(SemanticsFlag.isTextField), ifTrue: 'textField'));
+    final List<String> flags = SemanticsFlag.values.values.where((SemanticsFlag flag) => _hasFlag(flag)).map((SemanticsFlag flag) => flag.toString().substring('SemanticsFlag.'.length)).toList();
+    properties.add(new IterableProperty<String>('flags', flags, ifEmpty: null));
     properties.add(new FlagProperty('isInvisible', value: isInvisible, ifTrue: 'invisible'));
-    properties.add(new StringProperty('label', _label, defaultValue: ''));
-    properties.add(new StringProperty('value', _value, defaultValue: ''));
-    properties.add(new StringProperty('increasedValue', _increasedValue, defaultValue: ''));
-    properties.add(new StringProperty('decreasedValue', _decreasedValue, defaultValue: ''));
-    properties.add(new StringProperty('hint', _hint, defaultValue: ''));
+    properties.add(new StringProperty('label', _label.replaceAll('\n', '\\n'), defaultValue: ''));
```
Comment (jacob314) [inline packages/flutter/lib/src/semantics/semantics.dart]: You shouldn't have to do the replaceAll here. Please fix the text rendering code in DiagnosticsNode instead to handle \n in properties more gracefully. You should be able to change this in one spot in the large method implementing toStringDeep and get a benefit everywhere.

File: packages/flutter/lib/src/semantics/semantics.dart [hunk 2]
```diff
@@ -1378,21 +1378,14 @@ class SemanticsNode extends AbstractNode with DiagnosticableTreeMixin {
...
-      properties.add(new FlagProperty('isEnabled', value: _hasFlag(SemanticsFlag.isEnabled), ifFalse: 'disabled'));
-    if (_hasFlag(SemanticsFlag.hasCheckedState))
-      properties.add(new FlagProperty('isChecked', value: _hasFlag(SemanticsFlag.isChecked), ifTrue: 'checked', ifFalse: 'unchecked'));
-    properties.add(new FlagProperty('isInMutuallyExcusiveGroup', value: _hasFlag(SemanticsFlag.isInMutuallyExclusiveGroup), ifTrue: 'mutually-exclusive'));
-    properties.add(new FlagProperty('isSelected', value: _hasFlag(SemanticsFlag.isSelected), ifTrue: 'selected'));
-    properties.add(new FlagProperty('isFocused', value: _hasFlag(SemanticsFlag.isFocused), ifTrue: 'focused'));
-    properties.add(new FlagProperty('isButton', value: _hasFlag(SemanticsFlag.isButton), ifTrue: 'button'));
-    properties.add(new FlagProperty('isTextField', value: _hasFlag(SemanticsFlag.isTextField), ifTrue: 'textField'));
+    final List<String> flags = SemanticsFlag.values.values.where((SemanticsFlag flag) => _hasFlag(flag)).map((SemanticsFlag flag) => flag.toString().substring('SemanticsFlag.'.length)).toList();
+    properties.add(new IterableProperty<String>('flags', flags, ifEmpty: null));
```
Comment (jacob314) [inline packages/flutter/lib/src/semantics/semantics.dart]: glad to see you are using ifEmpty here!

**Review by jacob314 — COMMENTED** [review]
Let me know if you have any questions about how to fix toStringDeep to render this multi-line case in a less confusing manner.

File: packages/flutter/test/foundation/diagnostics_test.dart [hunk 3]
```diff
@@ -1491,8 +1491,8 @@ void main() {
       equalsIgnoringHashCodes(
         'TestTree#00000\n'
         '   objects:\n'
-        '     Rect.fromLTRB(0.0, 0.0, 20.0, 20.0)\n'
-        '     Color(0xffffffff)\n',
+        '       Rect.fromLTRB(0.0, 0.0, 20.0, 20.0)\n'
```
Comment (jacob314) [inline packages/flutter/test/foundation/diagnostics_test.dart]: looks like these are already indented 2. why are they getting indented 4?

File: packages/flutter/test/foundation/diagnostics_test.dart [hunk 4]
```diff
@@ -657,9 +657,9 @@ void main() {
         ' │ ╚═══════════\n'
         ' └─child node C: TestTree#00000\n'
         '     foo:\n'
-        '     multi\n'
-        '     line\n'
-        '     value!\n',
+        '       multi\n'
```
Comment (jacob314) [inline packages/flutter/test/foundation/diagnostics_test.dart]: I think this reads a lot better

File: packages/flutter/lib/src/foundation/diagnostics.dart [hunk 5]
```diff
@@ -883,6 +883,8 @@ abstract class DiagnosticsNode {
 
         builder.write(
             config.isNameOnOwnLine || description.contains('\n') ? '\n' : ' ');
+        if (description.contains('\n'))
+          builder.prefixOtherLines += '  ';
```
Comment (jacob314) [inline packages/flutter/lib/src/foundation/diagnostics.dart]: how does this look with the dense style?

File: packages/flutter/lib/src/painting/matrix_utils.dart [hunk 6]
```diff
@@ -255,7 +255,7 @@ class MatrixUtils {
 List<String> debugDescribeTransform(Matrix4 transform) {
   if (transform == null)
     return const <String>['null'];
-  final List<String> matrix = transform.toString().split('\n').map((String s) => '  $s').toList();
```
Comment (jacob314) [inline packages/flutter/lib/src/painting/matrix_utils.dart]: nice

File: packages/flutter/lib/src/foundation/diagnostics.dart (hunk 5 above)
Comment (goderbauer) [reply packages/flutter/lib/src/foundation/diagnostics.dart]: No change, as expected and documented by this test: https://github.com/goderbauer/flutter/blob/7887b12807397ddeb1890dbea30f6e84708e742a/packages/flutter/test/foundation/diagnostics_test.dart#L422

**Comment by jacob314** [comment]
lgtm

**Review by jacob314 — APPROVED** [review]

**Pull request merged**